### PR TITLE
hoc2023: AI calculation bug fix

### PR DESCRIPTION
**Pull request by breville**
This a proposed fix for a bug in which pressing "Start over" would attempt a `reduce` of an empty array in `calculateMinMax`.  The issue was that we'd cleared the `inputs`, causing a re-render, and then this calculation would be attempted.  

In this fix, we only do this calculation when we have new "good" and "bad" effects to score. 

Follow-up to https://github.com/code-dot-org/code-dot-org/pull/54732.

### CODE REVIEW

File: apps/src/dance/ai/DanceAiModal.tsx [hunk 1]
```diff
@@ -148,6 +148,58 @@ const DanceAiModal: React.FunctionComponent = () => {
...
+  // (ie, a SINGLE emoji association with a foreground/background palette combination),
+  // and a maximum total score
+  // (ie, the sum of ALL selected emoji's associations with a foreground/background palette combination).
+  // Used to normalize and scale the data for easier differentiation between results by the user.
+  const calculateMinMax = useCallback(
+    (useInputs: string[]) => {
+      // The minimum individual score is selected across all generated effects (bad and good).
+      const minIndividualScore = Array.from(
+        Array(BAD_GENERATED_RESULTS_COUNT + 1).keys()
+      ).reduce((accumulator: number, currentValue: number) => {
```
Comment (fisher-alice) [inline apps/src/dance/ai/DanceAiModal.tsx:183]: Nit: Maybe rename `accumulator` as `currentMin` and `currentValue` as `step`.

File: apps/src/dance/ai/DanceAiModal.tsx [hunk 2]
```diff
@@ -148,6 +148,58 @@ const DanceAiModal: React.FunctionComponent = () => {
...
+        const scores = getScores(useInputs, currentValue);
+        const min = Math.min(...scores);
```
Comment (fisher-alice) [inline apps/src/dance/ai/DanceAiModal.tsx:185]: @bencodeorg, why did you need to make a deep copy of `scores` here? Just curious.

File: apps/src/dance/ai/DanceAiModal.tsx [hunk 3]
```diff
@@ -165,13 +217,15 @@ const DanceAiModal: React.FunctionComponent = () => {
           ),
           goodEffect: JSON.parse(currentValue),
         };
+
+        minMaxAssociations.current = calculateMinMax(currentInputs);
```
Comment (fisher-alice) [inline apps/src/dance/ai/DanceAiModal.tsx:221]: When there is an existing field value and the user 'starts over' by clicking on the AI header emoji field, do you mind walking through what happens at that point? Are we still in `INITIAL` mode?

**Review by fisher-alice — APPROVED** [review]
Looks good! I pulled branch and works great.
Just left a couple non-blocking questions.

File: apps/src/dance/ai/DanceAiModal.tsx (hunk 2 above)
Comment (thomasoniii) [reply apps/src/dance/ai/DanceAiModal.tsx:185]: I don't think there's anything copying here— `Math.min` doesn't want an array as its arguments, so this is spreading out the array into a list of args into it.

File: apps/src/dance/ai/DanceAiModal.tsx [hunk 4]
```diff
@@ -148,6 +148,58 @@ const DanceAiModal: React.FunctionComponent = () => {
...
+      }, Infinity);
+
+      // By definition, the maximum total score must come from the "good" effect.
+      const goodEffectScores = getScores(
+        useInputs,
+        BAD_GENERATED_RESULTS_COUNT
+      );
+      const maxTotalScore = goodEffectScores.reduce(
+        (sum, score) => (sum += score)
+      );
```
Comment (thomasoniii) [inline apps/src/dance/ai/DanceAiModal.tsx:196]: minor quibble - this can be: `sum + score`

_technically_ it doesn't matter in this case, but stylistically I think it's clearer not to mutate the `sum` value that came in. There are linter rules that can gripe about this, and it keeps us in the habit of avoiding possible side effects.

File: apps/src/dance/ai/DanceAiModal.tsx [hunk 5]
```diff
@@ -148,6 +148,58 @@ const DanceAiModal: React.FunctionComponent = () => {
...
+  // Used to normalize and scale the data for easier differentiation between results by the user.
+  const calculateMinMax = useCallback(
+    (useInputs: string[]) => {
+      // The minimum individual score is selected across all generated effects (bad and good).
+      const minIndividualScore = Array.from(
+        Array(BAD_GENERATED_RESULTS_COUNT + 1).keys()
+      ).reduce((accumulator: number, currentValue: number) => {
+        const scores = getScores(useInputs, currentValue);
+        const min = Math.min(...scores);
+        return min < accumulator ? min : accumulator;
```
Comment (thomasoniii) [inline apps/src/dance/ai/DanceAiModal.tsx:186]: If I'm reading this right, the previous 3 lines can be shortened down to just:
```
Math.min(accumulator, ...getScores(useInputs, currentValue))
```

File: apps/src/dance/ai/DanceAiModal.tsx [hunk 6]
```diff
@@ -148,6 +148,58 @@ const DanceAiModal: React.FunctionComponent = () => {
...
+      return [0, 0, 0];
+    },
+    [getGeneratedEffect]
+  );
+
+  // Calculates the minimum individual score
+  // (ie, a SINGLE emoji association with a foreground/background palette combination),
+  // and a maximum total score
+  // (ie, the sum of ALL selected emoji's associations with a foreground/background palette combination).
+  // Used to normalize and scale the data for easier differentiation between results by the user.
```
Comment (thomasoniii) [inline apps/src/dance/ai/DanceAiModal.tsx:177]: Not blocking, but...

This `reduce` call feels excessively clever to me and like it's hiding the logic. I get that it's building a range and that javascript stupidly doesn't have one of those built in, but that initial definition is tough to read at first (at least for me). And then there's an Infinity, it's doing a `min` on the individual scores, and a comparison, and returning the final minimum value. The logic feels tough to parse out.

What about something like this instead?
```
const minIndividualScore = Math.min(              // give Math.min a list of all the possible scores and find the single low point
  ...Array(BAD_GENERATED_RESULTS_COUNT + 1)       // array of the appropriate length
    .fill(undefined)                              // fill it with undefined so we can map over it
    .map((_, step) => getScores(useInputs, step)) // map the step to the getScores value
    .flat()                                       // flatten out the sub arrays into a single array, which is consumed by the Math.min
)
```

**Review by thomasoniii — APPROVED** [review]
I added several stylistic suggestions on the `calculateMinMax` function, but nothing blocking. 

so, lgtm 🚀

File: apps/src/dance/ai/DanceAiModal.tsx (hunk 3 above)
Comment (breville) [reply apps/src/dance/ai/DanceAiModal.tsx:221]: In that case, we jump straight to `SELECT_INPUTS` mode, as seen in the handler [here](https://github.com/code-dot-org/code-dot-org/blob/3c59f629cec4fec427bfabad290ee6bdbd54d15a/apps/src/dance/ai/DanceAiModal.tsx#L290-L298).

Comment (fisher-alice) [reply apps/src/dance/ai/DanceAiModal.tsx:221]: Thanks!

**Review by sanchitmalhotra126 — APPROVED** [review]
+1 to Jim and Alice's comments, but LGTM to me from a bug fix perspective!

One general comment - looking at this again and understanding it better in this PR, I do think `minMaxAssociations` may belong in component in state, rather than a ref. I know we've been super careful about not causing additional re-renders, but given that the min max is essentially just data that tells the `DanceAiScore` how to render itself, logically it feels like component state that should influence rendering whenever it updates. As this component grows I think we'd just want to be careful that by using refs to avoid re-renders, we aren't hiding any unnecessary/complicated dependencies that are causing those extra renders in the first place. Of course, this isn't something we need to change right now, but just to wanted to mention while we were looking at this 🙂

**Comment by breville** [comment]
Thanks all for the excellent feedback.  Merging as-is to fix the known issue, but we might want to incorporate some of this into a future revision!

**Comment by thomasoniii** [comment]
Just adding a note here - one easy way to manifest this original bug was to select 3 emojis, and then click on one after hitting the `Generate` button.
